### ISSUE TITLE
ZTransducer#collectAllN pulls too much data

### ISSUE THREAD

**Issue by simpadjo** [zio/zio#3559]
Testcase: 

```
//crashes
Stream(1,2,3) ++ Stream.fail("X")).transduce(ZTransducer.collectAllN(3)).take(1).runCollect
```

Why it happens: 
`collectAllN` relies on `foldWeightedDecompose`. 
When weight reaches 3, it doesn't emit an element immediately but rather waits until weight becomes strictly bigger. 
I tried to change `foldWeightedDecompose` to terminate earlier but it breaks `collectAllSetN`.

Why it is important:
I'm working on `Transducer#race` family of combinators. 
Racing transducers in the presence of chunking could lead to very surprising results.
When a single transducer emits at unexpected moment, this makes things even worse.

It could be fixed by adding a flag to  `foldWeightedDecompose` whether it should terminate early.
Or alternatively, by stopping digging the hole and implementing `collectAllN` separately.

@iravid WDYT?

**Comment by iravid** [comment]
Ah I hit that weight == max issue at some point during the development and thought I fixed it. This must be another edge case.

In order of preference:
- Ignore this altogether because we don't guarantee to pull the absolute minimum required by an operator, and there are countless other scenarios in which it might not be possible to pull exactly what is needed;
- Try to fix foldWeightedDecompose (but not with a flag; we need to identify a way to adjust the algorithm);
- Re-implement collectAllN without the generalized combinators.

**Comment by simpadjo** [comment]
I used to think the same. But today I spent almost an hour trying to understand what's wrong with my testcase which looked like `stream.transduce(Transducer.collectAllN(2).choose(Transducer.collectAllN()3))`. 
See https://github.com/zio/zio/pull/3557/files .
Seems that good-enough guarantees don't compose well in this case.

Regarding `foldWeightedDecompose`, both ways of handling `weight = max` are reasonable.
When we know that we are done - no reason to pull more.
OTOH if the next element might have zero weight, we want to try to consume it as well. 
It happens in `collectToSetN`
